### PR TITLE
Gemma4 resizing per layer inputs

### LINKED ISSUE #45276: [gemma4] resize_token_embeddings does not effect to embed_tokens_per_layer or output_embeddings
### System Info

- `transformers` version: 5.5.0
- Platform: Linux-6.6.113+-x86_64-with-glibc2.35
- Python version: 3.12.13
- Huggingface_hub version: 1.8.0
- Safetensors version: 0.7.0
- Accelerate version: 1.13.0
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.10.0+cu128 (CUDA)
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: Tesla T4

(Google Colaboratory GPU)

### Who can help?

@zucchini-nlp @Cyrilvallez

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Reproduce the behavior:

```
from transformers import Gemma4ForConditionalGeneration
mdl = Gemma4ForConditionalGeneration.from_pretrained("google/gemma-4-E2B-it")
e = mdl.get_input_embeddings()
f = mdl.model.language_model.embed_tokens_per_layer
g = mdl.get_output_embeddings()
print(e.num_embeddings, f.num_embeddings, g.out_features)
assert e.num_embeddings == f.num_embeddings == g.out_features
e = mdl.resize_token_embeddings(e.num_embeddings + 1)
f = mdl.model.language_model.embed_tokens_per_layer
g = mdl.get_output_embeddings()
print(e.num_embeddings, f.num_embeddings, g.out_features)
assert e.num_embeddings == f.num_embeddings == g.out_features
```

### Expected behavior

All `e.num_embeddings`, `f.num_embeddings` and `g.out_features` should be increased to 262145.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/45276 and https://github.com/huggingface/transformers/issues/45335

In gemma4 per-layer inputs have to be resized as long as they aren't part of soft multimodal tokens

Repro for T5 gemma:
```python
from transformers import T5GemmaForConditionalGeneration

model = T5GemmaForConditionalGeneration.from_pretrained("harshaljanjani/tiny-t5gemma-test")
encoder = model.resize_token_embeddings(model.vocab_size + 10)
decoder = model.model.decoder.embed_tokens
head = model.get_output_embeddings()
print(encoder.weight.shape, decoder.weight.shape, head.weight.shape)
# LM head resize is reverted back because of tying. Decoder is never resized
>>> torch.Size([256010, 512]) torch.Size([256000, 512]) torch.Size([256000, 512])
```

Gemma3n has soft mm tokens, and the current state of 3N is not good. I see unused vocab entries in `mm_projection` 😢  and if when apply simply resizing to per-layer-input, we'll get even more unused entries

Could be done in the correct way if we filter out mm-tokens, but I'd prefer to leave it for now

### CODE REVIEW

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 1]
```diff
@@ -960,12 +960,6 @@ def __init__(self, config: Gemma3Config):
         self.lm_head = nn.Linear(config.text_config.hidden_size, config.text_config.vocab_size, bias=False)
         self.post_init()
 
-    def get_input_embeddings(self):
-        return self.model.get_input_embeddings()
-
-    def set_input_embeddings(self, value):
-        self.model.set_input_embeddings(value)
-
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma3/modeling_gemma3.py:968]: same as base class, so no need to override

File: src/transformers/models/t5gemma/modeling_t5gemma.py [hunk 2]
```diff
@@ -961,6 +961,11 @@ def __init__(self, config: T5GemmaConfig):
 
     def set_output_embeddings(self, new_embeddings):
         self.lm_head.out_proj = new_embeddings
+        # The tying happens from decoder to lm-head, but when resizing
+        # the resized embed is assigned only to the head. Then tying weights
+        # again reverts everything back. So we have to update decoder here
+        if self.config.tie_word_embeddings:
+            self.model.decoder.embed_tokens = new_embeddings
```
Comment (zucchini-nlp) [inline src/transformers/models/t5gemma/modeling_t5gemma.py]: not a fan of it tbh, but ig it's better than overriding `resize`

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -2155,6 +2155,9 @@ def test_resize_tokens_embeddings(self):
 
             # Check that the model can still do a forward pass successfully (every parameter should be resized)
             if not is_deepspeed_zero3_enabled():
+                # Input ids should be expanded to the new maximum size of the vocabulary
+                inputs_dict["input_ids"][:, -2] = new_model_vocab_size - 1
+
```
Comment (zucchini-nlp) [inline tests/test_modeling_common.py:2167]: if we had this, we'd know that gemma4 resize doesn't work well. Added now

File: src/transformers/models/t5gemma/modeling_t5gemma.py [hunk 4]
```diff
@@ -961,6 +961,12 @@ def __init__(self, config: T5GemmaConfig):
 
     def set_output_embeddings(self, new_embeddings):
         self.lm_head.out_proj = new_embeddings
+        # The tying happens from decoder to lm-head, but when resizing
+        # the resized embed is assigned only to the head. Then tying weights
+        # again reverts everything back. So we have to update decoder here
+        if self.config.tie_word_embeddings:
+            self.model.decoder.embed_tokens.weight = new_embeddings.weight
+            self.model.decoder.embed_tokens.num_embeddings = new_embeddings.weight.shape[0]
```
Comment (KoichiYasuoka) [inline src/transformers/models/t5gemma/modeling_t5gemma.py:969]: Oh it's nice idea, but don't you need the following?

```
        self.model.decoder.embed_tokens.embedding_dim = new_embeddings.weight.shape[1]
```

Comment (zucchini-nlp) [reply src/transformers/models/t5gemma/modeling_t5gemma.py:969]: hidden dim doesn't change, only the number of entries :)

**Comment by zucchini-nlp** [comment]
@bot /repo

**Review by ArthurZucker — APPROVED** [review]
SGTM, its a bit dirty to have to call add hook to module but model specific so fine b y me!
